### PR TITLE
Improved fuzzy matching with warning

### PR DESCRIPTION
**Pull request by Casper-Guo**
Added test cases to the fuzzy version of `get_event_by_name` and reimplemented some logic to avoid false positive warnings.

I recommend considering using [RapidFuzz](https://github.com/rapidfuzz/RapidFuzz?tab=readme-ov-file) over [thefuzz](https://github.com/seatgeek/thefuzz?tab=readme-ov-file) since it seems to be more widely adopted and actively maintained. The performance difference is not important for our use case. The API is identical so this only requires changing the requirements. I will put it in a separate PR if you like the idea.

### CODE REVIEW

File: fastf1/tests/test_events.py [hunk 1]
```diff
@@ -132,6 +132,46 @@ def test_event_schedule_get_by_name():
...
+        "United States West Grand Prix").EventName == "United States Grand Prix West"
+    assert schedule.get_event_by_name(
+        "US West").EventName == "United States Grand Prix West"
+
+    # tests for common inputs
+    schedule = fastf1.get_event_schedule(2024)
+    # assert schedule.get_event_by_name(
+    #     "Saudi").EventName == "Saudi Arabian Grand Prix"
+    # assert schedule.get_event_by_name(
+    #     "Saudi GP").EventName == "Saudi Arabian Grand Prix"
```
Comment (Casper-Guo) [inline fastf1/tests/test_events.py:154]: I would love to be able to handle these cases correctly since they are reasonable inputs. This does not seem possible due to interference from `Abu Dhabi Grand Prix`. At least we are emitting a warning to users now.

Comment (theOehrly) [reply fastf1/tests/test_events.py:154]: What if you try to handle this in a way similar to the current ``driver_color`` and ``team_color`` functions? Before doing fuzzy matching, you add a check for partial exact matches. That way, these might work as well?

File: fastf1/events.py [hunk 2]
```diff
@@ -400,9 +401,12 @@ def get_event(
         force_ergast: [Deprecated, use ``backend='ergast'``] Always use data
             from the ergast database to create the event schedule
 
-        strict_search: Match precisely the query, or default to
+        strict_search: This argument is planned for deprecation, use the
```
Comment (theOehrly) [inline fastf1/events.py:404]: I think we are deprecating ``strict_search`` right away? Then the argument is planned for removal, but it already is deprecated. So the wording should be changed to "is deprecated".

File: fastf1/events.py [hunk 3]
```diff
@@ -981,16 +1002,21 @@ def get_event_by_name(
...
-                strict_search=True)``
+                exact_match=True)``
                 will return the event for the British Grand Prix, whereas
-                ``.get_event_by_name("British", strict_search=True)``
+                ``.get_event_by_name("British", exact_match=True)``
                 will return ``None``
         """
-
         if strict_search:
+            warnings.warn(("This argument is planned for deprecation, "
```
Comment (theOehrly) [inline fastf1/events.py:1016]: You should name the argument in the warning text to avoid confusion. "'strict_search' is deprecated, use... "

**Review by theOehrly — CHANGES_REQUESTED** [review]

File: fastf1/events.py (hunk 3 above)
Comment (Casper-Guo) [reply fastf1/events.py:1016]: Definitely, forgot these comments are not for docs

**Comment by theOehrly** [comment]
> I recommend considering using [RapidFuzz](https://github.com/rapidfuzz/RapidFuzz?tab=readme-ov-file) over [thefuzz](https://github.com/seatgeek/thefuzz?tab=readme-ov-file) since it seems to be more widely adopted and actively maintained. The performance difference is not important for our use case. The API is identical so this only requires changing the requirements. I will put it in a separate PR if you like the idea.

I'm fine with switching to rapidfuzz if the (apparently minor) API differences aren't relevant here.

**Comment by theOehrly** [comment]
Also, the test failure on 3.8 minimum versions is completely unrelated. That's a CI problem. I'm fixed that just now. If you want, you can rebase onto master. But it's not really necessary, I think.

File: fastf1/tests/test_events.py (hunk 1 above)
Comment (Casper-Guo) [reply fastf1/tests/test_events.py]: I tried that at first but doing a partial ratio only would cause failures in the `United States Grand Prix` vs `United States Grand Prix West` edge case.

Solution now is to check for partial matches first and fall back to using the full ratio if:

- there is zero 100% match
- there are multiple 100% matches

**Comment by Casper-Guo** [comment]
> I'm fine with switching to rapidfuzz if the (apparently minor) API differences aren't relevant here.

I will PR for that once this is merged

File: fastf1/tests/test_events.py (hunk 1 above)
Comment (theOehrly) [reply fastf1/tests/test_events.py]: That seems like a good solution 👍

**Comment by theOehrly** [comment]
Test failures are caused by the changes here. The ``reference_laps_data`` fixture that is defined in ``conftest.py`` now returns data for a different event.
Somehow ``fastf1.get_session(2020, 'Italy', 'R')`` now returns the "Eifel Grand Prix". I probably shouldn't have used "Italy" as identifier in a season with three Italian grand prix. But the result should at least be one of those three. 
The desired result in the fixture is the "Italian Grand Prix" (Round 8). We can switch to an explicit ``get_session(2020, 8, 'R')`` there. But the matching problem still needs to be fixed.

**Comment by Casper-Guo** [comment]
That's super old. Italy should be 100% matched to all the Italian races that season and we tiebreak between those. I'll look into why we get a German race instead.

**Comment by Casper-Guo** [comment]
I see the EventSchedule dataframe's index doesn't start at 0, probably to accommodate testing.

This is inside the `fuzzy_match` function
```
print(self.iloc[[0, 1]])
```

Output:

```
RoundNumber  Country   Location                                  OfficialEventName  EventDate            EventName   EventFormat  ...    Session4               Session4Date     Session4DateUtc Session5               Session5Date     Session5DateUtc F1ApiSupport
2            1  Austria  Spielberg  FORMULA 1 ROLEX GROSSER PREIS VON ÖSTERREICH 2020 2020-07-05  Austrian Grand Prix  conventional  ...  Qualifying  2020-07-04 15:00:00+02:00 2020-07-04 13:00:00     Race  2020-07-05 15:10:00+02:00 2020-07-05 13:10:00         True
3            2  Austria  Spielberg  FORMULA 1 PIRELLI GROSSER PREIS DER STEIERMARK... 2020-07-12   Styrian Grand Prix  conventional  ...  Qualifying  2020-07-11 15:00:00+02:00 2020-07-11 13:00:00     Race  2020-07-12 15:10:00+02:00 2020-07-12 13:10:00         True
```

Is this expected?

**Review by theOehrly — APPROVED** [review]

**Comment by theOehrly** [comment]
> I see the EventSchedule dataframe's index doesn't start at 0, probably to accommodate testing.

This isn't to "accommodate" testing. But it is an artifact of the fact that testing exists in the schedule. ``include_testing=False`` filters the testing events out, but slicing in Pandas doesn't reindex the DataFrame. So the index starts at a non-zero value then. 
I thought a bit about whether this behaviour is desired, or whether ``.get_event_schedule`` should explicitly reindex the DataFrame before returning it. There certainly are arguments for both sides. But I think in terms of compatibility of the data with itself and for consistency within FastF1, filtering out the data and retaining the original index is better. For example, the  ``.pick_*`` methods on ``Laps`` don't reindex the data either. Additionally, this would potentially be a breaking change for some users. Therefore, I don't think we really want to change that behaviour.

If you don't have any further objections, I'll merge.

**Comment by Casper-Guo** [comment]
No more changes needed on my front. Please merge if it looks good to you

**Comment by theOehrly** [comment]
Great, this is overall a pretty nice improvement 👍

**Comment by Casper-Guo** [comment]
I will open another one to switch the fuzzing library
